app/scripts/metamask-controller.js

```diff
@@ -3036,6 +3089,117 @@ export default class MetamaskController extends EventEmitter {
+      async (currentValue, previousValue) => {
+        const changedAuthorizations = getChangedAuthorizations(
+          currentValue,
+          previousValue,
+        );
+
+        const removedAuthorizations = getRemovedAuthorizations(
+          currentValue,
+          previousValue,
+        );
+
+        // remove any existing notification subscriptions for removed authorizations
+        for (const [origin, authorization] of removedAuthorizations.entries()) {
+          const sessionScopes = getSessionScopes(authorization);
+          // if the eth_subscription notification is in the scope and eth_subscribe is in the methods
+          // then remove middleware and unsubscribe
+          Object.entries(sessionScopes).forEach(([scope, scopeObject]) => {
+            if (
+              scopeObject.notifications.includes('eth_subscription') &&
+              scopeObject.methods.includes('eth_subscribe')
+            ) {
+              this.multichainMiddlewareManager.removeMiddlewareByScopeAndOrigin(
+                scope,
+                origin,
+              );
+              this.multichainSubscriptionManager.unsubscribeByScopeAndOrigin(
+                scope,
+                origin,
+              );
+            }
+          });
+        }
+
+        // add new notification subscriptions for changed authorizations
+        for (const [origin, authorization] of changedAuthorizations.entries()) {
+          const sessionScopes = getSessionScopes(authorization);
+
+          // if the eth_subscription notification is in the scope and eth_subscribe is in the methods
+          // then get the subscriptionManager going for that scope
+          Object.entries(sessionScopes).forEach(([scope, scopeObject]) => {
+            if (
+              scopeObject.notifications.includes('eth_subscription') &&
+              scopeObject.methods.includes('eth_subscribe')
+            ) {
+              // for each tabId
+              Object.values(this.connections[origin]).forEach(({ tabId }) => {
+                const subscriptionManager =
+                  this.multichainSubscriptionManager.subscribe({
+                    scope,
+                    origin,
+                    tabId,
+                  });
+                this.multichainMiddlewareManager.addMiddleware({
+                  scope,
+                  origin,
+                  tabId,
+                  middleware: subscriptionManager.middleware,
+                });
+              });
+            } else {
+              this.multichainMiddlewareManager.removeMiddlewareByScopeAndOrigin(
+                scope,
+                origin,
+              );
+              this.multichainSubscriptionManager.unsubscribeByScopeAndOrigin(
+                scope,
+                origin,
+              );
+            }
+          });
+
+          // TODO: could be pushed into selectors?
+          const previousAuthorization = previousValue.get(origin);
```

Comment:
How is this case different than `removedAuthorizations`?

Reply:
removal of a permission entirely vs removal of a scope from a permission

Reply:
please see https://github.com/MetaMask/metamask-extension/pull/27782/commits/f90be7c21514022584d49641154192315036b5dd which combines this block of logic to determining when just a scope is removed with the existing helper that returns when an entire permission is removed/revoked